File: backends/go-service/internal/services/leaderboard.go
package services

import (
	"context"
	"fmt"
	"time"

	"github.com/yourusername/health-competition-go/internal/models"

	"github.com/redis/go-redis/v9"
)

type LeaderboardService struct {
	cache       *CacheService
	redisClient *redis.Client
}

func NewLeaderboardService(cache *CacheService, redisClient *redis.Client) *LeaderboardService {
	return &LeaderboardService{
		cache:       cache,
		redisClient: redisClient,
	}
}

// GetLeaderboard retrieves the leaderboard for a competition
func (s *LeaderboardService) GetLeaderboard(ctx context.Context, competitionID string, limit int) (*models.Leaderboard, error) {
	key := s.getLeaderboardKey(competitionID)

	// Get top N entries from sorted set (reverse order - highest scores first)
	entries, err := s.cache.ZRevRangeWithScores(ctx, key, 0, int64(limit-1))
	if err != nil {
		return nil, err
	}

	leaderboardEntries := make([]models.LeaderboardEntry, 0, len(entries))
	for i, entry := range entries {
		userID := entry.Member.(string)
		
		// Get user details from cache
		userDetails, err := s.getUserDetails(ctx, competitionID, userID)
		if err != nil {
			// If user details not found, create minimal entry
			userDetails = &models.LeaderboardEntry{
				UserID: userID,
				UserName: "Unknown",
			}
		}

		leaderboardEntry := models.LeaderboardEntry{
			UserID:        userID,
			UserName:      userDetails.UserName,
			CompetitionID: competitionID,
			Score:         int64(entry.Score),
			Rank:          i + 1,
			Steps:         userDetails.Steps,
			Distance:      userDetails.Distance,
			Calories:      userDetails.Calories,
			LastSyncedAt:  userDetails.LastSyncedAt,
			UpdatedAt:     time.Now(),
		}
		leaderboardEntries = append(leaderboardEntries, leaderboardEntry)
	}

	totalCount, err := s.redisClient.ZCard(ctx, key).Result()
	if err != nil {
		totalCount = int64(len(entries))
	}

	return &models.Leaderboard{
		CompetitionID: competitionID,
		Entries:       leaderboardEntries,
		TotalCount:    int(totalCount),
		UpdatedAt:     time.Now(),
	}, nil
}

// UpdateScore updates a user's score in the leaderboard
func (s *LeaderboardService) UpdateScore(ctx context.Context, req *models.ScoreUpdateRequest) error {
	key := s.getLeaderboardKey(req.CompetitionID)

	// Calculate total score (you can customize this formula)
	score := float64(req.Steps)

	// Update score in sorted set
	err := s.cache.ZAdd(ctx, key, score, req.UserID)
	if err != nil {
		return err
	}

	// Store detailed user data
	userDetailsKey := s.getUserDetailsKey(req.CompetitionID, req.UserID)
	userDetails := &models.LeaderboardEntry{
		UserID:        req.UserID,
		CompetitionID: req.CompetitionID,
		Score:         int64(score),
		Steps:         req.Steps,
		Distance:      req.Distance,
		Calories:      req.Calories,
		LastSyncedAt:  time.Now(),
		UpdatedAt:     time.Now(),
	}

	err = s.cache.Set(ctx, userDetailsKey, userDetails, 24*time.Hour)
	if err != nil {
		return err
	}

	// Publish update to Redis pub/sub for WebSocket broadcasting
	s.publishLeaderboardUpdate(ctx, req.CompetitionID, req.UserID, int64(score))

	return nil
}

// GetUserRank gets the rank of a specific user
func (s *LeaderboardService) GetUserRank(ctx context.Context, competitionID, userID string) (int, error) {
	key := s.getLeaderboardKey(competitionID)
	
	// Get reverse rank (0-based), add 1 to make it 1-based
	rank, err := s.cache.ZRevRank(ctx, key, userID)
	if err != nil {
		return 0, err
	}

	return int(rank) + 1, nil
}

// CalculatePrizes calculates prize distribution for a competition
func (s *LeaderboardService) CalculatePrizes(ctx context.Context, competitionID string, prizePool float64) ([]models.Prize, error) {
	// Get top 3 winners
	leaderboard, err := s.GetLeaderboard(ctx, competitionID, 3)
	if err != nil {
		return nil, err
	}

	if len(leaderboard.Entries) == 0 {
		return nil, fmt.Errorf("no participants in competition")
	}

	// Prize distribution: 60%, 30%, 10%
	distribution := models.PrizeDistribution{
		Rank1Percentage: 0.60,
		Rank2Percentage: 0.30,
		Rank3Percentage: 0.10,
	}

	prizes := make([]models.Prize, 0)

	// Rank 1
	if len(leaderboard.Entries) > 0 {
		prizes = append(prizes, models.Prize{
			ID:            fmt.Sprintf("prize-%s-1", competitionID),
			CompetitionID: competitionID,
			UserID:        leaderboard.Entries[0].UserID,
			Rank:          1,
			Amount:        prizePool * distribution.Rank1Percentage,
			Status:        "pending",
			CreatedAt:     time.Now(),
		})
	}

	// Rank 2
	if len(leaderboard.Entries) > 1 {
		prizes = append(prizes, models.Prize{
			ID:            fmt.Sprintf("prize-%s-2", competitionID),
			CompetitionID: competitionID,
			UserID:        leaderboard.Entries[1].UserID,
			Rank:          2,
			Amount:        prizePool * distribution.Rank2Percentage,
			Status:        "pending",
			CreatedAt:     time.Now(),
		})
	}

	// Rank 3
	if len(leaderboard.Entries) > 2 {
		prizes = append(prizes, models.Prize{
			ID:            fmt.Sprintf("prize-%s-3", competitionID),
			CompetitionID: competitionID,
			UserID:        leaderboard.Entries[2].UserID,
			Rank:          3,
			Amount:        prizePool * distribution.Rank3Percentage,
			Status:        "pending",
			CreatedAt:     time.Now(),
		})
	}

	// Cache prizes
	prizesKey := s.getPrizesKey(competitionID)
	s.cache.Set(ctx, prizesKey, prizes, 7*24*time.Hour)

	return prizes, nil
}

// Helper methods
func (s *LeaderboardService) getLeaderboardKey(competitionID string) string {
	return fmt.Sprintf("leaderboard:%s", competitionID)
}

func (s *LeaderboardService) getUserDetailsKey(competitionID, userID string) string {
	return fmt.Sprintf("user_details:%s:%s", competitionID, userID)
}

func (s *LeaderboardService) getPrizesKey(competitionID string) string {
	return fmt.Sprintf("prizes:%s", competitionID)
}

func (s *LeaderboardService) getUserDetails(ctx context.Context, competitionID, userID string) (*models.LeaderboardEntry, error) {
	key := s.getUserDetailsKey(competitionID, userID)
	var details models.LeaderboardEntry
	err := s.cache.Get(ctx, key, &details)
	if err != nil {
		return nil, err
	}
	return &details, nil
}

func (s *LeaderboardService) publishLeaderboardUpdate(ctx context.Context, competitionID, userID string, score int64) {
	channel := fmt.Sprintf("leaderboard:%s", competitionID)
	message := map[string]interface{}{
		"type":           "score_update",
		"competition_id": competitionID,
		"user_id":        userID,
		"score":          score,
		"timestamp":      time.Now(),
	}
	s.cache.Publish(ctx, channel, message)
}

File: backends/python-service/app/services/leaderboard.py
from datetime import datetime
from typing import List
from app.models.types import (
    Leaderboard,
    LeaderboardEntry,
    Prize,
    PrizeDistribution,
    ScoreUpdateRequest
)
from app.services.cache import CacheService

class LeaderboardService:
    def __init__(self, cache: CacheService):
        self.cache = cache

    async def get_leaderboard(self, competition_id: str, limit: int = 100) -> Leaderboard:
        """Get leaderboard for a competition"""
        key = f"leaderboard:{competition_id}"
        
        entries_data = await self.cache.zrevrange_withscores(key, 0, limit - 1)
        
        leaderboard_entries = []
        for i, (user_id, score) in enumerate(entries_data):
            user_details = await self._get_user_details(competition_id, user_id)
            
            if not user_details:
                user_details = {
                    "user_id": user_id,
                    "user_name": "Unknown",
                    "steps": 0,
                    "distance": 0.0,
                    "calories": 0.0,
                    "last_synced_at": datetime.utcnow().isoformat(),
                }
            
            entry = LeaderboardEntry(
                user_id=user_id,
                user_name=user_details.get("user_name", "Unknown"),
                competition_id=competition_id,
                score=int(score),
                rank=i + 1,
                steps=user_details.get("steps", 0),
                distance=user_details.get("distance", 0.0),
                calories=user_details.get("calories", 0.0),
                last_synced_at=datetime.fromisoformat(
                    user_details.get("last_synced_at", datetime.utcnow().isoformat())
                ),
                updated_at=datetime.utcnow()
            )
            leaderboard_entries.append(entry)
        
        total_count = await self.cache.zcard(key)
        
        return Leaderboard(
            competition_id=competition_id,
            entries=leaderboard_entries,
            total_count=total_count,
            updated_at=datetime.utcnow()
        )

    async def update_score(self, req: ScoreUpdateRequest) -> bool:
        """Update user score in leaderboard"""
        key = f"leaderboard:{req.competition_id}"
        score = float(req.steps)
        
        await self.cache.zadd(key, score, req.user_id)
        
        # Store detailed user data
        user_details_key = f"user_details:{req.competition_id}:{req.user_id}"
        user_details = {
            "user_id": req.user_id,
            "competition_id": req.competition_id,
            "score": int(score),
            "steps": req.steps,
            "distance": req.distance,
            "calories": req.calories,
            "last_synced_at": datetime.utcnow().isoformat(),
            "updated_at": datetime.utcnow().isoformat(),
        }
        
        await self.cache.set(user_details_key, user_details, 86400)
        
        # Publish update
        await self._publish_leaderboard_update(req.competition_id, req.user_id, int(score))
        
        return True

    async def calculate_prizes(
        self, competition_id: str, prize_pool: float
    ) -> List[Prize]:
        """Calculate prize distribution for a competition"""
        leaderboard = await self.get_leaderboard(competition_id, 3)
        
        if not leaderboard.entries:
            raise ValueError("No participants in competition")
        
        distribution = PrizeDistribution()
        prizes = []
        
        if len(leaderboard.entries) > 0:
            prizes.append(Prize(
                id=f"prize-{competition_id}-1",
                competition_id=competition_id,
                user_id=leaderboard.entries[0].user_id,
                rank=1,
                amount=prize_pool * distribution.rank_1_percentage,
                status="pending",
                created_at=datetime.utcnow()
            ))
        
        if len(leaderboard.entries) > 1:
            prizes.append(Prize(
                id=f"prize-{competition_id}-2",
                competition_id=competition_id,
                user_id=leaderboard.entries[1].user_id,
                rank=2,
                amount=prize_pool * distribution.rank_2_percentage,
                status="pending",
                created_at=datetime.utcnow()
            ))
        
        if len(leaderboard.entries) > 2:
            prizes.append(Prize(
                id=f"prize-{competition_id}-3",
                competition_id=competition_id,
                user_id=leaderboard.entries[2].user_id,
                rank=3,
                amount=prize_pool * distribution.rank_3_percentage,
                status="pending",
                created_at=datetime.utcnow()
            ))
        
        # Cache prizes
        prizes_key = f"prizes:{competition_id}"
        await self.cache.set(prizes_key, [p.dict() for p in prizes], 604800)
        
        return prizes

    async def _get_user_details(self, competition_id: str, user_id: str):
        """Get user details from cache"""
        key = f"user_details:{competition_id}:{user_id}"
        return await self.cache.get(key)

    async def _publish_leaderboard_update(
        self, competition_id: str, user_id: str, score: int
    ):
        """Publish leaderboard update to Redis pub/sub"""
        channel = f"leaderboard:{competition_id}"
        message = {
            "type": "score_update",
            "competition_id": competition_id,
            "user_id": user_id,
            "score": score,
            "timestamp": datetime.utcnow().isoformat(),
        }
        await self.cache.publish(channel, message)
